### ISSUE TITLE
Ability to generate "just" an expression, without a select

### ISSUE THREAD

**Issue by romandecker** [knex/knex#2526]
# Feature discussion / request

I would like to be able to generate only an expression with knex. I'm using [join-monster](https://www.npmjs.com/package/join-monster) for my GraphQL-postgres API, and in order to filter my data, I need to supply it with a WHERE-clause, and only a WHERE-clause ([see here for an example])(https://join-monster.readthedocs.io/en/latest/where/).

Now, generating the clause can be somewhat complex, because I may have to adjust it depending on the parameters supplied by the client. It would be super nice if I could harness knex for that - join-monster actually already recommends knex for query execution, so I have it in my project already.

However, as far as I can see, there is no way of creating **only** an expression with knex. I can only generate the clause as part of a select. Example:

    console.log(
      knex.queryBuilder()
      .where('a', 1)
      .andWhere('b', '>=', 77)
      .toString()
    );
    // -> "select * where \"a\" = 1 and \"b\" >= 77"
    //                    ^^^^^^^^^^^^^^^^^^^^^^^^^^
    //                        what I actually need

My current ugly workaround is to match out the expression I need:

    console.log(
      knex.queryBuilder()
      .where('a', 1)
      .andWhere('b', '>=', 77)
      .toString()
      .match(/where (.*)?/)[1]
    );
    // -> "\"a\" = 1 and \"b\" >= 77"

Which gives me what I need. But that seems like a very bad approach to me.

Is there a better way to get **just** the expression I need here?

**Comment by wubzz** [comment]
Unfortunately there's no way to do that in the public knex API.

The following very hacky example works, but of course: Don't use it.

```javascript
const queryBuilder = knex
    .where('a', 1)
    .andWhere('b', '>=', 77);

  const compiler = queryBuilder.client.queryCompiler(queryBuilder);
  const clause = compiler.where();

  console.log(queryBuilder.client._formatQuery(clause, compiler.formatter.bindings).substr(6));
//"a" = 1 and "b" >= 77
```

Overall I don't think it would be too difficult to implement.

A simple solution would be to not force-append the `SELECT * ` statement when calling `.toString()` *(changing this would yield the same result as my hack above)*. Breaking change, of course.

**Comment by wubzz** [comment]
This is actually very simply to add:

```javascript
//src/query/builder.js
 toClauseString() {
    this._method = 'where';

    return (this.toQuery() || '').substr(6);
  },
```

Some questions are:
1) Better name for function?
2) Is this something we want to add to knex?

What do you think @elhigu @tgriesser ?

**Comment by elhigu** [comment]
Yes, I've been wanting to have this so many times! however there is one problem. We cannot use toQuery() nor .toString() for generating it, because those are vulnerable to SQL injection.

So rendering method should be something like `builder.toClauseSQL()` which returns `knex.raw` 
 object like this: 

```js
{ 
  sql: '"a" = ? and "b" >= ?', 
  bindings:[1, 77]
}
```

Which can be then used for example like this:

```js
knex('table').where(whereClause)
```

or 

```js
knex.raw('select * from ?? where ?', ['MyTable', whereClause])
```

It would be also useful to be able to build sql for other separate parts of query, like select and joins etc. but where part I've missed the most.

**Comment by wubzz** [comment]
Yeah that should be easy enough.

```javascript
//src/query/builder.js
 toWhereClauseSQL() {
    this._method = 'where';

    const result = this.toSQL();

    result.sql = result.sql.substr(6);

    return result;
  },
```

Untested, but should yield expected result.

**Comment by elhigu** [comment]
@wubzz if I also wouldn't like query builder internal to be changed when that is called, so maybe:

```js
 toWhereClauseSQL() {
    const whereBuilder = this.clone();

    whereBuilder._method = 'where';

    const result = whereBuilder.toSQL();

    result.sql = result.sql.substr(6);

    return result;
  },
```

**Comment by mercmobily** [comment]
Thanks for the discussion on this feature request. After careful consideration, I don't think it's a good idea to add `toWhereClauseSQL()` and similar methods to Knex core. I am going through all of the issues, and I am afraid this request shouldn't go through.

While the use case is valid (generating SQL fragments for tools like join-monster), adding these methods would:

- Expand the API surface for a relatively niche use case
- Open the door to requests for every other clause type (`toSelectClauseSQL()`, `toJoinClauseSQL()`, etc.)
- Shift Knex's focus from being a query builder to an expression builder

For those needing this functionality, there are good alternatives:
1. Use `knex.raw()` with bindings for simple expressions
2. Extract the WHERE clause from the generated SQL when needed
3. Consider a dedicated expression builder library for complex cases

We appreciate the thoughtful PR from @wubzz and the valuable input from @elhigu on this. The discussion helped clarify Knex's scope and boundaries.

If this becomes a widely-requested feature in the future, we can revisit. For now, we'll keep Knex focused on building complete queries rather than fragments.
